python/ray/rllib/agents/dqn/distributional_q_model.py

```diff
@@ -0,0 +1,228 @@
+                    action_out,
+                    units=self.action_space.n * num_atoms,
+                    activation=None)
+            else:
+                action_scores = model_outputs
+            if num_atoms > 1:
+                # Distributional Q-learning uses a discrete support z
+                # to represent the action value distribution
+                z = tf.range(num_atoms, dtype=tf.float32)
+                z = v_min + z * (v_max - v_min) / float(num_atoms - 1)
+                support_logits_per_action = tf.reshape(
+                    tensor=action_scores,
+                    shape=(-1, self.action_space.n, num_atoms))
+                support_prob_per_action = tf.nn.softmax(
+                    logits=support_logits_per_action)
+                action_scores = tf.reduce_sum(
+                    input_tensor=z * support_prob_per_action, axis=-1)
+                logits = support_logits_per_action
+                dist = support_prob_per_action
+                return [
+                    action_scores, z, support_logits_per_action, logits, dist
+                ]
+            else:
+                logits = tf.expand_dims(tf.ones_like(action_scores), -1)
+                dist = tf.expand_dims(tf.ones_like(action_scores), -1)
+                return [action_scores, logits, dist]
+
+        def build_state_score(model_out):
+            state_out = model_out
+            for i in range(len(q_hiddens)):
+                if use_noisy:
+                    state_out = self._noisy_layer("dueling_hidden_%d" % i,
+                                                  state_out, q_hiddens[i],
+                                                  sigma0)
+                elif parameter_noise:
+                    state_out = tf.contrib.layers.fully_connected(
+                        state_out,
+                        num_outputs=q_hiddens[i],
+                        activation_fn=tf.nn.relu,
+                        normalizer_fn=tf.contrib.layers.layer_norm)
+                else:
+                    state_out = tf.layers.dense(
+                        state_out, units=q_hiddens[i], activation=tf.nn.relu)
+            if use_noisy:
+                state_score = self._noisy_layer(
+                    "dueling_output",
+                    state_out,
+                    num_atoms,
+                    sigma0,
+                    non_linear=False)
+            else:
+                state_score = tf.layers.dense(
+                    state_out, units=num_atoms, activation=None)
+            return state_score
+
+        def build_action_value_in_scope(model_out):
+            with tf.variable_scope(
+                    name + "/action_value", reuse=tf.AUTO_REUSE):
+                return build_action_value(model_out)
+
+        def build_state_score_in_scope(model_out):
+            with tf.variable_scope(name + "/state_value", reuse=tf.AUTO_REUSE):
+                return build_state_score(model_out)
```

Comment:
This is probably a super bad pattern, but I'm not sure how to better port the old code without a rewrite.